packages/ui/src/components/multi-select-v2.tsx

```diff
@@ -0,0 +1,329 @@
+  (
+    {
+      label,
+      caption,
+      value,
+      onChange,
+      placeholder,
+      defaultValue = [],
+      options: arrayOptions,
+      searchQuery,
+      setSearchQuery,
+      disabled,
+      className,
+      disallowCreation = false,
+      commandProps,
+      inputProps
+    }: MultiSelectProps,
+    ref: React.Ref<MultiSelectRef>
+  ) => {
+    const inputRef = useRef<HTMLInputElement>(null)
+    const [open, setOpen] = useState(false)
+    const [onScrollbar, setOnScrollbar] = useState(false)
+    const dropdownRef = useRef<HTMLDivElement>(null) // Added this
+
+    const [selected, setSelected] = useState<MultiSelectOption[]>(value || defaultValue || [])
+    const [options, setOptions] = useState<MultiSelectOption[]>(arrayOptions || [])
+    const [inputValue, setInputValue] = useState('')
+    const { search } = useDebounceSearch({
+      handleChangeSearchValue: setSearchQuery,
+      searchValue: searchQuery || ''
+    })
+
+    const isControlled = !!value
+
+    useImperativeHandle(
+      ref,
+      () => ({
+        selectedValue: [...selected],
+        input: inputRef.current as HTMLInputElement,
+        focus: () => inputRef?.current?.focus(),
+        reset: () => setSelected([])
+      }),
+      [selected]
+    )
+
+    const handleClickOutside = (event: MouseEvent | TouchEvent) => {
+      if (
+        dropdownRef.current &&
+        !dropdownRef.current.contains(event.target as Node) &&
+        inputRef.current &&
+        !inputRef.current.contains(event.target as Node)
+      ) {
+        setOpen(false)
+        inputRef.current.blur()
+      }
+    }
+
+    const handleUnselect = useCallback(
+      (option: MultiSelectOption) => {
+        if (isControlled) {
+          const newOptions = value?.filter(s => s.id !== option.id) || []
+          onChange?.(newOptions)
+          option.onReset?.()
+        } else {
+          const newOptions = selected.filter(s => s.id !== option.id)
+          setSelected(newOptions)
+          option.onReset?.()
+        }
+      },
+      [onChange, selected, isControlled, value]
+    )
+
+    const handleKeyDown = useCallback(
+      (e: React.KeyboardEvent<HTMLDivElement>) => {
+        const input = inputRef.current
+        if (input) {
+          if (e.key === 'Delete' || e.key === 'Backspace') {
+            if (input.value === '' && (isControlled ? value?.length : selected.length) > 0) {
+              handleUnselect((isControlled ? value : selected)[(isControlled ? value : selected).length - 1])
+            }
+          }
+          if (e.key === 'Enter' && input.value && !disallowCreation) {
+            if (
+              !options.some(option => option.key.toLowerCase() === input.value.toLowerCase()) &&
+              !(isControlled ? value : selected).some(s => s.key.toLowerCase() === input.value.toLowerCase())
+            ) {
+              let newOption: MultiSelectOption
+
+              if (input.value.includes(':')) {
```

Comment:
we can pull this out into a util function which returns early as conditions match